### PR TITLE
feat: add tracing to request

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/context.go [hunk 1]
```diff
@@ -421,3 +423,61 @@ func buildRequestContext(w http.ResponseWriter, r *http.Request, opContext *oper
...
+
+func ParseRequestTraceOptions(r *http.Request, disableTracing bool) (options resolve.RequestTraceOptions) {
+	if disableTracing {
+		options.DisableAll()
+		return
+	}
+	var (
+		values []string
+	)
+	if r.Header.Get(RequestTraceHeader) != "" {
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/context.go:448]: Get uses `textproto.MIMEHeader()` under the hood which your method doesn't. I'd stay with the current approach.

File: router/core/context.go [hunk 2]
```diff
@@ -421,3 +423,61 @@ func buildRequestContext(w http.ResponseWriter, r *http.Request, opContext *oper
 	}
 	return requestContext
 }
+
+const (
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/context.go:427]: yes

File: router/core/context.go [hunk 3]
```diff
@@ -421,3 +423,61 @@ func buildRequestContext(w http.ResponseWriter, r *http.Request, opContext *oper
...
+	// RequestTraceHeader is the header used to enable request tracing
+	RequestTraceHeader = "X-WG-Trace"
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/context.go:429]: ok

File: router/config/config.go [hunk 4]
```diff
@@ -159,8 +159,9 @@ type EngineDebugConfiguration struct {
 }
 
 type EngineExecutionConfiguration struct {
-	Debug              EngineDebugConfiguration
-	EnableSingleFlight bool `default:"true" envconfig:"ENGINE_ENABLE_SINGLE_FLIGHT"`
+	Debug                 EngineDebugConfiguration
+	EnableSingleFlight    bool `default:"true" envconfig:"ENGINE_ENABLE_SINGLE_FLIGHT"`
+	DisableRequestTracing bool `default:"false" envconfig:"ENGINE_DISABLE_REQUEST_TRACING"`
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/config/config.go:164]: ok

File: router/core/graphql_prehandler.go [hunk 5]
```diff
@@ -55,12 +60,19 @@ func (h *PreHandler) Handler(next http.Handler) http.Handler {
...
 		// In GraphQL the statusCode does not always express the error state of the request
 		// we use this flag to determine if we have an error for the request metrics
-		var hasRequestError bool
-		statusCode := http.StatusOK
-		var writtenBytes int
+		var (
+			hasRequestError bool
+			writtenBytes    int
+			statusCode      = http.StatusOK
+			traceOptions    = ParseRequestTraceOptions(r, h.disableRequestTracing)
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/graphql_prehandler.go:67]: ok

File: router/core/operation_planner.go [hunk 6]
```diff
@@ -64,14 +65,28 @@ func (p *OperationPlanner) preparePlan(requestOperationName []byte, requestOpera
...
+		opType:       operation.Type,
+		content:      operation.NormalizedRepresentation,
+		hash:         operation.ID,
+		clientInfo:   clientInfo,
+		variables:    operation.Variables,
+		traceOptions: traceOptions,
+	}
+
+	if traceOptions.Enable {
+		// if we have tracing enabled we always prepare a new plan
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/operation_planner.go:81]: ok

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**